text/3730-nit-lint-level.md

```diff
@@ -0,0 +1,379 @@
+```console
+$ cargo clippy
+   Compiling foo
+    Finished `foo` profile [optimized + debuginfo] target(s) in 15.85s
+```
+No lints reported.
+
+A non-rust-analyzer user could run the following to look for feedback:
+```console
+$ CARGO_BUILD_NITS=nit cargo clippy
+   Compiling foo
+note: creating a let-binding and then immediately returning it like `let x = expr; x` at the end of a block
+...
+    Finished `foo` profile [optimized + debuginfo] target(s) in 15.85s
+```
+
+They then go and post a PR.
+On the PR, the code above gets the following issue opened:
+```
+X Check failure
+
+Code scanning clippy
+
+creating a let-binding and then immediately returning it like `let x = expr; x` at the end of a block (warning)
+
+Show more details
+```
+This gives the contributor an opportunity to know of this potential improvement
+if they missed it and for the code reviewer to double check if they agreed with
+the author's determination to ignore it.
+
+For future PRs, Github should not report this as it should recognize that the report is for existing code.
+
+## Choosing lint levels
+
+When creating a lint or overriding a default lint level,
+its important to consider the semantics, something like:
+
+`error` is for hard errors that even prevent reporting of errors further in the build, running of tests, etc.
+This should be reserved for when there is no point building or testing anything further because of how broken the code is.
+
+`warn` generally has the semantics of a soft-error.
+Its non-blocking locally but will be blocked when the code is merged upstream.
+This should be used when eventual correctness or consistency is desired.
+
+`nit` is for coding suggestions for users that may or may not improve the code or may not need to be done yet.
+
+`allow` is for when there is no right answer (e.g. mutually exclusive lints), there are false positives, the lint is overly noisy, or the lint is of limited value.
+
+*Note:* This is for communicating the intent of this RFC and to get people started with this new feature and is not meant to precisely redefine the
+[lint levels](https://rustc-dev-guide.rust-lang.org/diagnostics.html#diagnostic-levels).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Rustc
+
+Rustc focuses on the mechanics of a new, visible lint level `nit`.
+
+Add support for the new lint:
+```rust
+#![nit(deprecated)]
+```
+```console
+$ rustc --nit deprecated ...
+```
+```console
+$ rustc -Ndeprecated ...
+```
+
+These will be rendered like other lint levels with a diagnostic level and coloring that conveys the non-blocking, helpful nature.
+For example, this could be rendered like a `note:` or `help:` diagnostic level.
+
+Rustc will have a dynamic lint group of all `nit`s, much like `warnings` is all lints at level `warn` at the time of processing, `-Anits`.
+
+## Cargo
+
+Cargo implements the semantics for this to be a first-class non-blocking lint level.
```

Comment:
@Lokathor from https://github.com/rust-lang/rfcs/pull/3730#discussion_r1848861769

> doesn't that just give nit fatigue?
> 
> Like fundamentally if there's a lot of code you're in charge of, and it's doing something we don't detect, then we detect it, you'll have a lot to fix. and it doesn't matter if we call that new check a nit or a warning. if it suddenly says there's 4000 "whatever" in your project, you'll sigh deeply.

Reply:
epage from https://github.com/rust-lang/rfcs/pull/3730#discussion_r1848890656

> The intention of this is that maintainers don't feel the need to know how many "whatever"s they have in a project. Ideally, both the code author (though LSP) and the code reviewer (through code review integration) see the feedback and decide whether it should be acted on. Or put another way, these are only of interest for new code and can be ignored otherwise.
> 
> The one exception is if you are using this to adopt a new lint gradually. In that case you can either fix every report for a lint of interest in which case you only care about the numbers for that specific lint going down or you only worry about new code, assuming the existing code is battle tested enough and through evolution will eventually be rewritten in which case you still don't care about the metric.

Reply:
@Lokathor from https://github.com/rust-lang/rfcs/pull/3730#discussion_r1848979859

> I understand your intent, but absolutely nothing in this RFC actually does that. Nowhere does this RFC actually explain how "new code" is automatically detected, and nits only run against that. Are you assuming that it only scans for nits in the current git diff, or something like that? Such a system needs to be explained in the RFC if you're hanging the design on it.
> 
>     * What about people not using git?
> 
>     * what about people not using any revision control at all? does this never show up in the rust playground since that doesn't use version control?

Reply:
In the Guide section, the lints are shown to that user in two primary scenarios
- Rust-analyzer.  I'm not sure what is capable here and so I framed that unknown as a drawback.  It might not be on "only new code".  Hopefully it can be done in a way that is not too noisy for users.
- A CI integration posting to the code review.  I left this vague because not everyone has the same stack and technologies come and go.  I did mention `clippy-sarif` which is what I am using in my own projects.  [Github tracks these across runs](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#providing-data-to-track-code-scanning-alerts-across-runs) and I was given the impression that it will only report for new alerts but not finding it in the docs atm

Reply:
Yeah, I think the easy part of this RFC is saying "show me diagnostics on new code only". I think the hard part is to define a system for what's "new code". If you can't actually say how cargo or rustc or rust-analyzer is picking out new code and only showing diagnostics for that, then the RFC is only half done.

At the same time, if there's a system to run a lint against "new code only", as a lesser level than "warn", then any time rustc adds a new lint people can just set it the lint to "new_only" if they feel it's too much of a change to their codebase all at once. And all existing lints that got allowed through can be bumped to new and then bumped to warn eventually.

So I'd suggest focusing on what "new code" means, and trying to define that more specifically.

Reply:
I've summarized the Github side of my comment in 9e3c40a and reached out to r-a people in https://github.com/rust-lang/rfcs/pull/3730#discussion_r1849048869